Move tag query evaluation logic into tag expressions

replay opened the pull request:
Moves the tag query expression evaluation logic into the expression types. This provides us more flexibility to combine expressions from meta records and tag queries when we have to take both indexes into account.

One notable change to the logic is that MetricDefinition filters now assume that the tag index they're currently looking at isn't necessarily the only tag index. 
F.e. if an expression says `tag1!=abc` and we see that `metric1` does not have `tag1` defined via its metric (intrinsic) tags, then we can't be sure yet whether another index (f.e. meta tag index) does assign `tag1=abc` to `metric1`. So in this case the filter would return a `None` decision, which means that then other index(es) would need to get checked for whether this expression is satisfied or not. 
On the other hand if the metric (intrinsic) tags would assign `tag1=abc` to `metric1`, then it would directly return a conclusive decision `Fail` and sub-subsequent index(es) can be ignored. If the metric (intrinsic) tags would assign `tag1=otherValue` to `metric1`, then it would directly return a conclusive decision `Pass` because the indexes are checked in order of their priority of conflict resolution.

Currently the lookups in the meta tag index are not implemented yet, this will be coming in the next PR.

The current status of this PR is that it works fine in my manual tests and the unit tests pass. I'm going to add more benchmarks and potentially add some performance improvements, but I'm not expecting any major changes.

This implements step `2)` of this list: https://github.com/grafana/metrictank/issues/660#issuecomment-490490846

UPDATE, trying to explain the above stuff a bit better:

Filter Functions (Matchers):
The given query expressions, once parsed into the types implementing `tagquery.Expression`, now provide a method called `GetMetricDefinitionFilter()`. This method returns a filter function (in other contexts this is also called a "matcher") which can be passed a MetricDefinition and it will return a decision which is either `Pass`, `Fail` or `None` (inconclusive). This allows us to build a chain of filter functions, first from the given expressions, but also from the query expressions associated with meta records which match the given expressions. 
So let's say we have an expression `tag=~val.*` and want to apply it as a filter (not to build the initial result set). Once it's parsed into the according `tagquery.Expression` we call `GetMetricDefinitionFilter()` in `TagQueryContext.prepareExpressions()` to obtain a filter function which we'll use to filter down the result set. 
Here comes the important part -> 
Additionally, once we want to use the meta tag index, we can use the same `tagquery.Expression` instance to lookup meta tag records (`tagquery.MetaTagRecord`) which match the expression `tag=~val.*` from the meta tag index (`UnpartitionedMemoryIdx.metaTags`). When we do that lookup from the meta tag index we use the `tagquery.Expression` instance not as a filter, but to build a set of meta tag records which match the expression `tag=~val.*`. We do that via its methods `GetKey()` and `ValuePasses()` (because the `=~` operator matches against the tag value of a key).
The resulting meta tag records come with one or multiple sets of query expressions (`MetaTagRecord.Expressions`). If any of these sets of query expressions match a `MetricDefinition` this means that the meta tags associated with the meta tag record get assigned to that Metric and we already know that one of the meta tags matches the expression `tag=~val.*`, which means that the expression `tag=~val.*` matches this metric (via the meta tag that gets assigned to it). 
Concretely, to implement this, `prepareExpressions()` will lookup those meta tag records as described and build a set of filter chains from them (because there are multiple sets of expressions associated with each meta tag record, and one expression can result in multiple meta tag records). These filter chains can then be used by `testByAllExpressions()`. If a MetricDefinition passes any one of the meta tag record based filter chains then it should be part of the result set (so there's a logical OR there).
There will probably be some tuning regarding which filter chain is cheaper to evaluate and which is more expensive, we should try to filter the result set down using the cheap filters first before applying the expensive ones, that's why `tagquery.Expression` has a `GetCostMultiplier()` method, to estimate cost of evaluation.

Filter Decisions:
The metric definition filters (`tagquery.MetricDefinitionFilter`) return a filter decision (`tagquery.FilterDecision`) which can be one of `Fail`, `Pass` or `None`. The `Fail` and `Pass` value simply mean that a filter has been able to conclusively decide that a given `MetricDefinition` has either passed the filter or has been disqualified by it. The `None` case is necessary because in certain cases a filter cannot make this decision without also looking at the other indexes that need to be taken into account. Let's take the example `tag!=value`:
Once we use the meta tag index, the function `testByAllExpressions()` will first apply a filter function that's been generated from the expression `tag!=value` to each `MetricDefinition` of the result set. If a `MetricDefinition` comes with a tag `tag=value`, the filter can directly return `Fail`. If the `MetricDefinition` comes with a tag `tag=otherValue`, the filter can directly return `Pass`. But if the `MetricDefinition` does not have the tag `tag` defined at all, then this filter can't make a final decision because it is possible that one of the meta tag records assigns the tag `tag=value` to the given `MetricDefinition`, so it returns `None`. In the case of `None`, `testByAllExpressions()` will have to run the given `MetricDefinition` through the filter chains that have been generated from the meta tag records assigning the tag `tag=value` (as described above). If one of them matches the given `MetricDefinition`, then it can be disqualified at that point and we can interrupt the filter chains. If none of them match the given `MetricDefinition`, the `tagquery.Expression` object's `GetDefaultDecision()` is used to make the final decision whether the `MetricDefinition` matches this expression or not. In the case of the type `tagquery.expressionNotEqual` (representing `!=`) the return value of `GetDefaultDecision()` would be `Pass`. In the case of other expression types, like for example `tagquery.expressionEqual` (`=`) the return value of `GetDefaultDecision()` would be `Fail`. Because if a tag does not get assigned to a metric and we're evaluating a `=` expression, then the decision should be `Fail`. If a tag does not get assigned to a metric and we're evaluating a `!=` expression, then the result should be `Pass`.

replay commented:
Just FYI, those are the current benchmark results compared between master and this branch:
```
replay@mst-nb:~$  benchcmp /tmp/bench_master  /tmp/bench_branch
benchmark                                                        old ns/op     new ns/op     delta
BenchmarkTagQueryFilterAndIntersect/partitioned-8                11478558      40059410      +248.99%
BenchmarkTagQueryFilterAndIntersect/unPartitioned-8              47054050      179546200     +281.57%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/partitioned-8       31461560      44589916      +41.73%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/unPartitioned-8     156438270     200434630     +28.12%
BenchmarkTagQueryKeysByPrefixSimple/partitioned-8                15807         15468         -2.14%
BenchmarkTagQueryKeysByPrefixSimple/unPartitioned-8              2261          2217          -1.95%
BenchmarkTagQueryKeysByPrefixExpressions/partitioned-8           19747003      9970316       -49.51%
BenchmarkTagQueryKeysByPrefixExpressions/unPartitioned-8         66596885      10764884      -83.84%
BenchmarkTagQueryFilterByEqualExpression/partitioned-8           9365065       9606763       +2.58%
BenchmarkTagQueryFilterByEqualExpression/unPartitioned-8         28489946      34461254      +20.96%
BenchmarkTagQueryFilterByMatchExpression/partitioned-8           49737850      42749714      -14.05%
BenchmarkTagQueryFilterByMatchExpression/unPartitioned-8         297312400     244258900     -17.84%
BenchmarkTagQueryFilterByHasTagExpression/partitioned-8          62335045      57052060      -8.48%
BenchmarkTagQueryFilterByHasTagExpression/unPartitioned-8        359227266     335862066     -6.50%

benchmark                                                        old allocs     new allocs     delta
BenchmarkTagQueryFilterAndIntersect/partitioned-8                7690           5233           -31.95%
BenchmarkTagQueryFilterAndIntersect/unPartitioned-8              3338           883            -73.55%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/partitioned-8       235127         34635          -85.27%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/unPartitioned-8     205174         4759           -97.68%
BenchmarkTagQueryKeysByPrefixSimple/partitioned-8                32             32             +0.00%
BenchmarkTagQueryKeysByPrefixSimple/unPartitioned-8              4              4              +0.00%
BenchmarkTagQueryKeysByPrefixExpressions/partitioned-8           26992          983            -96.36%
BenchmarkTagQueryKeysByPrefixExpressions/unPartitioned-8         3460           206            -94.05%
BenchmarkTagQueryFilterByEqualExpression/partitioned-8           17050          17009          -0.24%
BenchmarkTagQueryFilterByEqualExpression/unPartitioned-8         16615          16607          -0.05%
BenchmarkTagQueryFilterByMatchExpression/partitioned-8           370968         34930          -90.58%
BenchmarkTagQueryFilterByMatchExpression/unPartitioned-8         369369         33379          -90.96%
BenchmarkTagQueryFilterByHasTagExpression/partitioned-8          501437         164938         -67.11%
BenchmarkTagQueryFilterByHasTagExpression/unPartitioned-8        500794         164792         -67.09%

benchmark                                                        old bytes     new bytes     delta
BenchmarkTagQueryFilterAndIntersect/partitioned-8                709176        511772        -27.84%
BenchmarkTagQueryFilterAndIntersect/unPartitioned-8              310965        131580        -57.69%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/partitioned-8       18413345      2381103       -87.07%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/unPartitioned-8     16453477      419914        -97.45%
BenchmarkTagQueryKeysByPrefixSimple/partitioned-8                2624          2624          +0.00%
BenchmarkTagQueryKeysByPrefixSimple/unPartitioned-8              352           352           +0.00%
BenchmarkTagQueryKeysByPrefixExpressions/partitioned-8           1704966       211470        -87.60%
BenchmarkTagQueryKeysByPrefixExpressions/unPartitioned-8         240589        36368         -84.88%
BenchmarkTagQueryFilterByEqualExpression/partitioned-8           4691539       4687239       -0.09%
BenchmarkTagQueryFilterByEqualExpression/unPartitioned-8         4068719       4067700       -0.03%
BenchmarkTagQueryFilterByMatchExpression/partitioned-8           36166214      9278016       -74.35%
BenchmarkTagQueryFilterByMatchExpression/unPartitioned-8         35042364      8164630       -76.70%
BenchmarkTagQueryFilterByHasTagExpression/partitioned-8          68448358      41549945      -39.30%
BenchmarkTagQueryFilterByHasTagExpression/unPartitioned-8        63688786      36831418      -42.17%
```
Most of them are looking good, but two show a significant slow-down. I'll investigate where that slow down comes from and try to fix it.

replay commented:
The good news is that I now have a good understand of why some benchmarks are performing worse, the bad news is that I have no good idea how to improve that. 

- In the current master we only have one tag index. This means that to evaluate the condition `tag=value` we only needs to check whether a certain id exists in the tag index under the keys `[tag][value]`. If it does the condition is satisfied, otherwise it is not.
- We now need to work based on the assumption that there are multiple indexes that can assign tags to a metric, with a defined order of precedence. This means when we see that a metric does not exist in `[tag][value]`, then we need to iterate over its tags and check if any of them assign the tag `tag` with a different value to it. If that's the case then we can directly return saying that the condition is not satisfied, because we check the different tag indexes in decreasing order of precedence. Only if by iterating over a metric's `.Tags` property we don't see the tag that we're looking for assigned with a different value, then we go on to check the next index.

The difference in these benchmarks comes from the additional looping over the `.Tags` property if there was no exact match for `tag=value`, to check whether there's another entry beginning with `tag=`.

replay commented:
I think to solve this above issue with the performance regression I'll add another option similar to `tag-support = true/false`, I'll call it `meta-tag-support = true/false`. That way it should be possible to avoid performance regressions at least when meta tag support is turned off.

replay commented:
With the latest 3 commits the performance has gotten better, but not as good as master yet:
```
benchmark                                                        old ns/op     new ns/op     delta
BenchmarkTagQueryFilterAndIntersect/partitioned-8                40323084      31360866      -22.23%
BenchmarkTagQueryFilterAndIntersect/unPartitioned-8              180773119     149006891     -17.57%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/partitioned-8       45430404      45239445      -0.42%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/unPartitioned-8     199925766     195143375     -2.39%
BenchmarkTagQueryKeysByPrefixSimple/partitioned-8                17038         17804         +4.50%
BenchmarkTagQueryKeysByPrefixSimple/unPartitioned-8              2291          2328          +1.62%
BenchmarkTagQueryKeysByPrefixExpressions/partitioned-8           9962729       9969944       +0.07%
BenchmarkTagQueryKeysByPrefixExpressions/unPartitioned-8         10184006      10596547      +4.05%
BenchmarkTagQueryFilterByEqualExpression/partitioned-8           11629725      11567934      -0.53%
BenchmarkTagQueryFilterByEqualExpression/unPartitioned-8         34105115      29431685      -13.70%
BenchmarkTagQueryFilterByMatchExpression/partitioned-8           47443844      47445716      +0.00%
BenchmarkTagQueryFilterByMatchExpression/unPartitioned-8         238119214     239374546     +0.53%
BenchmarkTagQueryFilterByHasTagExpression/partitioned-8          72438808      74312474      +2.59%
BenchmarkTagQueryFilterByHasTagExpression/unPartitioned-8        317500930     323824766     +1.99%

benchmark                                                        old allocs     new allocs     delta
BenchmarkTagQueryFilterAndIntersect/partitioned-8                5159           5123           -0.70%
BenchmarkTagQueryFilterAndIntersect/unPartitioned-8              825            822            -0.36%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/partitioned-8       34615          34623          +0.02%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/unPartitioned-8     4758           4759           +0.02%
BenchmarkTagQueryKeysByPrefixSimple/partitioned-8                32             32             +0.00%
BenchmarkTagQueryKeysByPrefixSimple/unPartitioned-8              4              4              +0.00%
BenchmarkTagQueryKeysByPrefixExpressions/partitioned-8           984            991            +0.71%
BenchmarkTagQueryKeysByPrefixExpressions/unPartitioned-8         206            207            +0.49%
BenchmarkTagQueryFilterByEqualExpression/partitioned-8           17009          17000          -0.05%
BenchmarkTagQueryFilterByEqualExpression/unPartitioned-8         16610          16611          +0.01%
BenchmarkTagQueryFilterByMatchExpression/partitioned-8           34920          34927          +0.02%
BenchmarkTagQueryFilterByMatchExpression/unPartitioned-8         33355          33357          +0.01%
BenchmarkTagQueryFilterByHasTagExpression/partitioned-8          164903         164917         +0.01%
BenchmarkTagQueryFilterByHasTagExpression/unPartitioned-8        164780         164776         -0.00%

benchmark                                                        old bytes     new bytes     delta
BenchmarkTagQueryFilterAndIntersect/partitioned-8                496563        496810        +0.05%
BenchmarkTagQueryFilterAndIntersect/unPartitioned-8              104305        101565        -2.63%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/partitioned-8       2357742       2358837       +0.05%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/unPartitioned-8     395076        388860        -1.57%
BenchmarkTagQueryKeysByPrefixSimple/partitioned-8                2624          2624          +0.00%
BenchmarkTagQueryKeysByPrefixSimple/unPartitioned-8              352           352           +0.00%
BenchmarkTagQueryKeysByPrefixExpressions/partitioned-8           210184        210344        +0.08%
BenchmarkTagQueryKeysByPrefixExpressions/unPartitioned-8         33915         33963         +0.14%
BenchmarkTagQueryFilterByEqualExpression/partitioned-8           4687368       4687182       -0.00%
BenchmarkTagQueryFilterByEqualExpression/unPartitioned-8         4068257       4068524       +0.01%
BenchmarkTagQueryFilterByMatchExpression/partitioned-8           9274441       9274677       +0.00%
BenchmarkTagQueryFilterByMatchExpression/unPartitioned-8         8120561       8120910       +0.00%
BenchmarkTagQueryFilterByHasTagExpression/partitioned-8          41543443      41544661      +0.00%
BenchmarkTagQueryFilterByHasTagExpression/unPartitioned-8        36818300      36818546      +0.00%
```

Dieterbe commented:
>iterating over a metric's .Tags property

This should be really fast.  Do we do more than merely iterate and check for equality ? Any parsing or more logic going on ?

replay commented:
@Dieterbe no, it just used to compare the strings for equality, while now it is doing a lookup by id from a map: https://github.com/grafana/metrictank/pull/1373/commits/39e35e2fcaffbf08498159d3a149f84c259bb514#diff-cb68716cfded2a2336a3778f548d7a17R50
keep in mind that within one query this can happen many millions of times, so an insignificant seeming difference can have impact

replay commented:
With the latest push things have improved a lot. I discovered today that another reason why the benchmarks in this working branch were showing worse performance was because the sorting by expression cost in this branch was not as good anymore as in master. So I fixed that, now we have better sorting again and together with the other improvements that I added the benchmarks actually look really good now.

This is comparing master with the latest commit in this branch (`37b3.fa`), running those benchmarks which previously have shown a worse performance:

```
replay@mst-nb:~/go/src/github.com/grafana/metrictank/idx/memory$ benchcmp /tmp/bench.d4047399 /tmp/bench.37b3faf6
benchmark                                                        old ns/op     new ns/op     delta
BenchmarkTagQueryFilterAndIntersect/partitioned-8                15076606      9971341       -33.86%
BenchmarkTagQueryFilterAndIntersect/unPartitioned-8              49369377      11758999      -76.18%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/partitioned-8       40507835      30111017      -25.67%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/unPartitioned-8     145715884     156101314     +7.13%

benchmark                                                        old allocs     new allocs     delta
BenchmarkTagQueryFilterAndIntersect/partitioned-8                7623           587            -92.30%
BenchmarkTagQueryFilterAndIntersect/unPartitioned-8              3282           248            -92.44%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/partitioned-8       235135         25169          -89.30%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/unPartitioned-8     205170         4114           -97.99%

benchmark                                                        old bytes     new bytes     delta
BenchmarkTagQueryFilterAndIntersect/partitioned-8                699451        218019        -68.83%
BenchmarkTagQueryFilterAndIntersect/unPartitioned-8              298730        62456         -79.09%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/partitioned-8       18416190      1819865       -90.12%
BenchmarkTagQueryFilterAndIntersectOnlyRegex/unPartitioned-8     16430234      350143        -97.87%
```

replay commented:
FYI rebased onto master

Dieterbe commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 # master
 
 ## breaking changes
+* version v0.12.0-182-ged2adc5b and later move the option `memory-idx.match-cache-size` to the new config section `tag-query.match-cache-size`.
```
why can't we keep this in memory-idx, because this option is a setting affecting the behavior of queries on the memory index, no?

Dieterbe replied:
similar for meta-tag-support, this seems to belong in memory-idx section just like tag-support already is.

replay replied:
Simply because that's where they are used when the query expressions get evaluated. It's definitely possible to leave both of them in the memory index, so the setting is there, then the `tagquery` package will just have to import the `idx.memory` package to access it.
Although, I just realized that in the next PR i'd probably have to move the `meta-tag-support` to `idx.memory` anyway, so we might as well do this now.

Dieterbe replied:
yes, our internal package layout is not a concern that should leak to operators of metrictank

Dieterbe commented on idx/memory/memory.go:
```diff
@@ -148,79 +163,179 @@ func (ids IdSet) String() string {
-type defByTagSet map[uint32]map[string]map[*schema.MetricDefinition]struct{}
+func (t TagIndex) idHasTag(id schema.MKey, tag, value uintptr) bool {
+	_, ok := t[tag][value][id]
+	return ok
```
this can panic if one of the intermediate maps doesn't exist.
maybe we should look up the metricdef by id ?

replay replied:
i don't think this should panic because when a non-existent key is looked up the return value is `nil`. `nil` is a valid map for reading (although not for writing).
https://play.golang.org/p/ZaYEovPqMuF

Dieterbe replied:
huh. whoa!

Dieterbe commented on expr/tagquery/meta_tag_record.go:
```diff
@@ -33,13 +33,13 @@ func ParseMetaTagRecord(metaTags []string, queries []string) (MetaTagRecord, err
 // It is assumed that all the queries are already sorted
-func (m *MetaTagRecord) MatchesQueries(other MetaTagRecord) bool {
-	if len(m.Queries) != len(other.Queries) {
+func (m *MetaTagRecord) MatchesQueries(other *MetaTagRecord) bool {
```
KISS. just call this `Equals` and make the comments less redundant.

replay replied:
Good idea

replay replied:
Considering that this only compares the Expressions of the meta tag record, I think it would probably be better to name the method `EqualExpressions`. Otherwise that would be confusing if it's called `Equal` but it only compares a part of the meta tag record.

Dieterbe commented on expr/tagquery/expression.go:
```diff
@@ -45,134 +37,336 @@ func (e Expressions) Strings() []string {
+	return nil, fmt.Errorf("ParseExpression: Invalid operator in expression %s", expr)
 }
 
+func ExpressionsAreEqual(expr1, expr2 Expression) bool {
```
why not make an Expression.Equals method?

replay replied:
Agree, that's better

Dieterbe commented on idx/memory/meta_tags.go:
```diff
@@ -75,9 +75,8 @@ func (m metaTagRecords) upsert(record tagquery.MetaTagRecord) (recordId, *tagque
 
 // hashMetaTagRecord generates a hash of all the queries in the record
 func (m *metaTagRecords) hashMetaTagRecord(record tagquery.MetaTagRecord) recordId {
```
why is this a method of MetaTagRecords and not of tagquery.MetaTagRecord ?

replay replied:
good point, that would be better

Dieterbe commented on idx/memory/meta_tags_test.go:
```diff
@@ -380,3 +262,26 @@ func TestDeletingMetaRecord(t *testing.T) {
 	}
 }
+
+func metaTagRecordsAreEqual(record1, record2 *tagquery.MetaTagRecord) bool {
```
we should really just standardize on `<type>.Equals` methods?

replay replied:
Yeah, I think we'll actually need two of those `.Equals` methods on the meta tag record type. One is comparing the whole meta tag record, the other is comparing only the expressions. 
So I'll name them `.Equals` and `.EqualExpressions`.

Dieterbe commented on expr/tagquery/expression.go:
```diff
@@ -45,134 +37,336 @@ func (e Expressions) Strings() []string {
+	// GetKey returns tag to who's values this expression get's applied if it operates on the value
+	// (OperatorsOnTag returns "false")
+	// example:
+	// in the expression "tag1=value" GetKey() would return "tag1" and OperatesOnTag() returns "false"
```
this is confusing. i checked all the implementations and this always returns the tag key, wether or not OperatesOnTag is true or false.

replay replied:
true, i think this comment is old and out of date. I'll update it

Dieterbe commented on idx/memory/tag_query.go:
```diff
@@ -2,504 +2,240 @@ package memory
+		keyPtr, found := expr.GetKeyPtr()
+		if !found {
+			// the key of this expression has not been interned, this means it cannot exist in the index.
+			// so this expression will be cheap to evaluate
```
example? maybe as a comment

replay replied:
added

Dieterbe replied:
thanks. the new comment makes a lot of sense.
are there any expression types for which this optimization is not true?

replay replied:
In all cases where an expression's `.OperatesOnTag()` returns `false`, the return value of `.GetKey()` has to match the tag value in the index exactly. So if the return value of `.GetKey()` cannot be found in the interning layer, then we know that this tag is not present in the index.

Dieterbe commented on idx/memory/tag_query.go:
```diff
@@ -2,504 +2,240 @@ package memory
+			valuePtr, found := expr.GetValuePtr()
+			if !found {
+				// the key of this expression has not been interned, this means it cannot exist in the index.
+				// so this expression will be cheap to evaluate
```
example? maybe as a comment

replay replied:
added

Dieterbe replied:
when it says "the key" is that a typo for the tag value?
also is this optimization true for all expression types?

replay replied:
ah, right. copy paste mistake

Dieterbe commented on idx/memory/tag_query.go:
```diff
@@ -2,504 +2,240 @@ package memory
-func (q *TagQueryContext) getInitialByMatch(expr kvRe, idCh chan schema.MKey, stopCh chan struct{}) {
-	defer q.wg.Done()
+	q.filters = make([]tagquery.MetricDefinitionFilter, len(q.query.Expressions)-1)
+	q.defaultDecisions = make([]tagquery.FilterDecision, len(q.query.Expressions)-1)
```
why -1 ? as comment please

replay replied:
added comment

Dieterbe commented on idx/memory/tag_query.go:
```diff
@@ -2,504 +2,240 @@ package memory
+	i := 0
+	for _, cost := range costs {
+		if q.startWith < 0 && q.query.Expressions[cost.expressionIdx].RequiresNonEmptyValue() {
+			q.startWith = cost.expressionIdx
```
what's going on here? are we just setting the startWith to the first expression that requires a non empty value?
what if we don't find one? (and startWith remains < 0) ?

replay replied:
the expressions have at that point already been sorted by cost. so yes, we just take the first one that requires a non-empty value. 
every set of expressions must include at least one expression requiring a non-empty value according to https://graphite.readthedocs.io/en/latest/tags.html#querying
`tagquery.NewQuery` enforces this rule, here we trust that the tag queries which get passed into the index have already been validated. If somebody instantiates a tag query without using `tagquery.NewQuery` then it's their responsibility to ensure it is valid. I'll add a comment mentioning this.

Dieterbe commented on idx/memory/tag_query.go:
```diff
@@ -2,504 +2,240 @@ package memory
+		if !found {
+			// the key value has never been interned by the index, so it is safe to
+			// assume it does not exist. this means this query must result in an
+			// empty result set
```
why is the outcome so different from above where we said "so this expression will be cheap to evaluate"

replay replied:
When building the initial result set we try to look up the metric definitions that match the expression which has chosen to be the one we start with. If we see that the key of this expression does not even exist in the index (it can't because it has not been interned) then we don't need to look anything up, because we can conclude directly that the result set will be empty.
The filter expressions above which are "cheap to evaluate" act according to the same logic. If a filter check whether some `metricX` has `tag=value`, but it already knows the value `tag` does not exist in the index at all because it's not been interned, then it doesn't even need to look at any of the tags or the tag index to conclude that `metricX` does not have `tag=value`.

Dieterbe commented on idx/memory/tag_query.go:
```diff
@@ -2,504 +2,240 @@ package memory
+			valuePtr, found := expr.GetValuePtr()
+			if !found {
+				// the key value has never been interned by the index, so it is safe to
+				// assume it does not exist. this means this query must result in an
```
why is the outcome so different from above where we said "so this expression will be cheap to evaluate"

replay replied:
same explanation as above

Dieterbe commented:
so there seems to be 2 ways to match a MetricDefinition against an Expression:
1) call MetricDefinitionFilter and run the function on the MD
2) call OperatesOnTag() to get the name of the tag that the Expression looks at, and call ValuePasses() or ValueMatchesExactly() with the value of the tag (but what if OperatesOnTag returned false, does that mean we should pass the tag key? that seems pointless since we already know what the key is of the tag it looks at)

> We do that via its methods GetKey() and ValuePasses() (because the =~ operator matches against the tag value of a key).

can you point me to the code where this happens. the only calls of GetKey are see are in the Equals methods, in Expressions.Sort and in the unit tests.

replay commented:
yes, the `MetricDefinition` filter type is used to filter down metric definitions based on their `id`/`def.Name`/`def.Tags` attributes.
Actually `id` is functionally not necessary for that, it is used because in certain cases it is faster to just look up if the given ID exists in the tag index under a specific tag/value than to iterate over the provided tags slice, so that's a performance optimizations.

The `OperatesOnTag()` is mostly relevant when we build the initial result set in `getInitialIds()` (https://github.com/grafana/metrictank/blob/7f6192ccaf3dc3e5c00396de5481b614730e1bf8/idx/memory/tag_query.go#L113). Certain expressions are making their decisions by looking at the value of a specified tag, such as `=` / `=~` / `!=` / `!=~`. While others are making their decisions by only looking at the tag and ignoring the tag value, such as `__tag=~` / `__tag^=`. 
If the chosen initial expressions operates on the tag, then we pass the known tags into its `.ValuePasses()` here: https://github.com/grafana/metrictank/blob/7f6192ccaf3dc3e5c00396de5481b614730e1bf8/idx/memory/tag_query.go#L205
If the chosen initial expression operates on the value, then we look up the available values for the expression's tag (returned by `GetKey()`) and then decide by passing the tag value in `.ValuePasses()` here: https://github.com/grafana/metrictank/blob/7f6192ccaf3dc3e5c00396de5481b614730e1bf8/idx/memory/tag_query.go#L170

The text snippet that you were citing is referring to how we'll be able to use `.GetKey()` and `.GetValue()` once we implement the metric lookup from the meta tag index, which will be done in the next PR, but that's not ready to review yet. I guess I should have articulated that better:
```
Additionally, once we want to use the meta tag index, we can use the same tagquery.Expression
instance to lookup meta tag records (tagquery.MetaTagRecord) which match the expression
tag=~val.* from the meta tag index (UnpartitionedMemoryIdx.metaTags). To do that lookup from
the meta tag index we will use the tagquery.Expression instance not as a filter, but to build a set
of meta tag records which match the expression tag=~val.*. We will do that via its methods
GetKey() and ValuePasses() (because the =~ operator matches against the tag value of a key).
```

replay commented:
That's an attempt to show what's happening when a tag query gets evaluated. This reflects the current state of this branch, I'm going to extend it for meta tag lookups in the next PR.
![tag query evaluation](https://user-images.githubusercontent.com/195371/62332696-9cc4a700-b48e-11e9-8a22-0ea3519140a8.jpg)

replay commented:
also pushed a few more comments to explain the expression type

replay commented:
With the latest pushes I had to update that diagram a bit:
![tag query evaluation (1)](https://user-images.githubusercontent.com/195371/62507603-01576d00-b7d3-11e9-9643-d296a7467432.jpg)

replay commented:
I have rebased this onto a master without interning. I will update the above flow charts, because they are now not up2date anymore

replay commented:
This is an updated version of the flow chart:
![tag query evaluation](https://user-images.githubusercontent.com/195371/62566615-5fc62f00-b857-11e9-8c85-a0a47a442c08.jpg)

robert-milan commented on expr/tagquery/expression.go:
```diff
@@ -45,134 +40,346 @@ func (e Expressions) Strings() []string {
+	// GetOperatorCost returns a value which should roughly reflect the cost of this operator compared
+	// to other operators. F.e. = is cheaper than =~. Keep in mind that this is only a very rough
+	// estimate and will never be accurate.
+	GetOperatorCost() uint32
```
Is there a way that we could run benchmarks on different operators to better determine their cost and fine tune these? Or possibly add metrics about how long it takes to process certain operators so in the future we could do this?

replay replied:
Yeah, but the exact number isn't really very relevant as it's only used to compare the expression types against each other. I have done a bunch of benchmarking and tried to tune it as good as possible, which resulted in this commit: https://github.com/grafana/metrictank/pull/1373/commits/bc6c8b7748a7f6a4c35707b79f82ff5bd9dc2785
Basically the sorting by cost first orders by expression type, according to the expressions "operator costs", and then among equal costs it sorts by cardinality of the affected parts of the index

robert-milan commented on expr/tagquery/expression.go:
```diff
@@ -45,134 +40,346 @@ func (e Expressions) Strings() []string {
+	err := validateQueryExpressionTagKey(resCommon.key)
 	if err != nil {
-		return res, fmt.Errorf("Error when validating key \"%s\" of expression \"%s\": %s", res.Key, expression, err)
+		return nil, fmt.Errorf("Error when validating key \"%s\" of expression \"%s\": %s", resCommon.key, expr, err)
```
We should a prefix to this and all other errors contained inside `expr` and `tagquery` like we do with the indexes, such as `expr-tagquery: ` or `tagquery: `. This will allow us to more easily filter logs when looking for problems.

replay replied:
I was under the impression that we're only prefixing log statements with prefixes such as `expr-tagquery:`, because this helps in understanding the logs. Are we doing the same for returned error values which get returned to the user?

robert-milan replied:
Good point, no I don't think we are.

replay commented on expr/tagquery/expression.go:
```diff
@@ -45,134 +40,346 @@ func (e Expressions) Strings() []string {
+	valuePos := pos
+	for ; pos < len(expr); pos++ {
+		// disallow ; in value
+		if expr[pos] == 59 {
```
not really, replacing it with the char is fine, thx

replay commented on idx/memory/memory.go:
```diff
@@ -84,6 +82,8 @@ func ConfigSetup() {
 	memoryIdx.DurationVar(&findCacheBackoffTime, "find-cache-backoff-time", time.Minute, "amount of time to disable the findCache when the invalidate queue fills up.")
 	memoryIdx.StringVar(&indexRulesFile, "rules-file", "/etc/metrictank/index-rules.conf", "path to index-rules.conf file")
 	memoryIdx.StringVar(&maxPruneLockTimeStr, "max-prune-lock-time", "100ms", "Maximum duration each second a prune job can lock the index.")
+	memoryIdx.IntVar(&tagquery.MatchCacheSize, "match-cache-size", 1000, "size of regular expression cache in tag query evaluation")
```
actually that's what i've done at first, then @dieter asked me to keep the settings in the `memory-idx` (which i think makes sense). It's just unfortunate that now we have to set this setting across packages
https://github.com/grafana/metrictank/pull/1373#discussion_r308388681

robert-milan replied:
Is there something else we can do here then? Like keep them completely contained within the index and then after they are loaded initialize them into tagquery with its own variables to hold them? Although then if we ever want to change them at runtime it would be a bit trickier.

It just feels like we are creating a lot of codependency here.

replay replied:
Unfortunately the package `tagquery` can't import the package `memory`, this would result in a cycle. We could initialize a setting in `memory`, and then in `ConfigProcess()` basically just copy that value into `tagquery.MetaTagSupport` / `tagquery.MatchCacheSize`. Then both of these two packages would hold a copy of the setting.

robert-milan replied:
Yeah something like that.

replay replied:
done: https://github.com/grafana/metrictank/pull/1373/commits/aefbc3c0696294448ffc0501b3da0138fc44c354

replay commented on expr/tagquery/expression.go:
```diff
@@ -45,134 +40,346 @@ func (e Expressions) Strings() []string {
 }
 
+// MetricDefinitionFilter takes a metric name together with its tags and returns a FilterDecision
+type MetricDefinitionFilter func(id schema.MKey, name string, tags []string) FilterDecision
```
actually i have been reading the MT guidelines and I think they are a bit confusing:
```
When defining multiple functions or structures stick to this ordering:
- first the main structure, then other structures that it references (unless they belong in a different file or package of course)
- first the high-level function, then any other function that it calls (unless it belongs in a different file or package of course)
- first a structure, then its constructor, and then its methods
```
so if you have a `struct1` with a bunch of methods, and `struct2` is a member of `struct1`, should it be like:
```
type struct1 struct {
    struct2
}
func NewStruct1() *struct1
func (s *struct1) func1()
func (s *struct1) func1()

type struct2 struct {}
func NewStruct2() *struct2
func (s *struct2) func()
```
Or should it be like
```
type struct1 struct {
    struct2
}
type struct2 struct {}

func NewStruct1() *struct1
func (s *struct1) func1()
func (s *struct1) func1()

func NewStruct2() *struct2
func (s *struct2) func()
```

Dieterbe replied:
or option 3, which is
```
type struct1 struct {
    struct2
}
type struct2 struct {}

func NewStruct2() *struct2
func (s *struct2) func()

func NewStruct1() *struct1
func (s *struct1) func1()
func (s *struct1) func2()


```
if the declaration of struct2 and its methods is quite/very short compared to struct1, i would go with option 3.

if they are similarly verbose then i would do option 1.
if the declaration of struct2 is short, but its methods are verbose, then I guess your option 2 makes sense.

it's not always black/white, and we we can decide specifically to each situation.

if anyone has examples of better guidelines or of code (especially stdlib which i like to follow) that applies something different, i would be glad to see it.

replay replied:
In that case I actually like the current ordering of this file. It keeps the important types and their methods close to each other near the top, at the bottom there are only a few auxiliary types and their methods which are used by the types / funcs above them
